consensus/event_loop.go

```diff
@@ -0,0 +1,908 @@
+func (fc *fastHotStuffEventLoop) ProcessValidatorVote(vote VoteMessage) error {
+	// Grab the event loop's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the event loop is running. This guarantees that the chain tip and validator list
+	// have already been set.
+	if fc.status != eventLoopStatusRunning {
+		return errors.New("FastHotStuffEventLoop.ProcessValidatorVote: Event loop is not running")
+	}
+
+	// Do a basic integrity check on the vote message
+	if !isProperlyFormedVote(vote) {
+		return errors.New("FastHotStuffEventLoop.ProcessValidatorVote: Malformed vote message")
+	}
+
+	// Check if the vote is stale
+	if isStaleView(fc.currentView, vote.GetView()) {
+		return errors.Errorf("FastHotStuffEventLoop.ProcessValidatorVote: Vote has a stale view %d", vote.GetView())
+	}
+
+	// Check if the public key has already voted for this view. The protocol does not allow
+	// a validator to vote for more than one block in a given view.
+	if fc.hasVotedForView(vote.GetPublicKey(), vote.GetView()) {
+		return errors.Errorf(
+			"FastHotStuffEventLoop.ProcessValidatorVote: validator %s has already voted for view %d",
+			vote.GetPublicKey().ToString(),
+			vote.GetView(),
+		)
+	}
+
+	// Check if the public key has already timed out for this view. The protocol does not allow
+	// for a validator to vote for a block in a view that it has already timed out for.
+	if fc.hasTimedOutForView(vote.GetPublicKey(), vote.GetView()) {
+		return errors.Errorf(
+			"FastHotStuffEventLoop.ProcessValidatorVote: validator %s has already timed out for view %d",
+			vote.GetPublicKey().ToString(),
+			vote.GetView(),
+		)
+	}
+
+	// Compute the value sha3-256(vote.View, vote.BlockHash)
+	voteSignaturePayload := GetVoteSignaturePayload(vote.GetView(), vote.GetBlockHash())
+
+	// Verify the vote signature
+	if !isValidSignatureSinglePublicKey(vote.GetPublicKey(), vote.GetSignature(), voteSignaturePayload[:]) {
+		return errors.New("FastHotStuffEventLoop.ProcessValidatorVote: Invalid signature")
+	}
+
+	// Cache the vote in case we need it for later
+	fc.storeVote(voteSignaturePayload, vote)
+
+	// Check if the crank timer has elapsed or the event loop has constructed a QC in the current view.
+	// If so, then there's nothing more to do.
+	if !fc.hasCrankTimerRunForCurrentView || fc.hasConstructedQCInCurrentView {
+		return nil
+	}
+
+	// Check if the vote is for the chain tip. If not, then there's nothing more to do.
+	if vote.GetBlockHash() != fc.tip.block.GetBlockHash() {
+		return nil
+	}
+
+	// Check if we have a super-majority vote for the chain tip.
+	if voteQCEvent := fc.tryConstructVoteQCInCurrentView(); voteQCEvent != nil {
+		// Signal the server that we can construct a QC for the chain tip, and mark that we have
+		// constructed a QC for the current view.
+		fc.hasConstructedQCInCurrentView = true
+		fc.Events <- voteQCEvent
```

Comment:
So one thing to note about this queuing here is that IF we ever hit the limit of 10,000 events on the channel, then I think we're guaranteed to deadlock. This is because one function will be adding to this queue while holding the lock, but no other function can run until the lock is released.

I think to avoid this you have to release the lock before you enqueue. But this is also kindof messy and annoying. Also I think that it won't even work because server.go is synchronous.

The simpler solution is to just call this in a Goroutine. That allows the function to return and release the lock so that other things can dequeue. I think this will cause the "limit" you placed on the channel to be effectively useless, though, since you'll queue up potentially unlimited goroutines. It's probably still better than deadlocking though.

Reply:
Your suggestion is definitely better than having deadlock risk. I've addressed this in https://github.com/deso-protocol/core/pull/836

Although in practice, if the event loop has a backlog of 10K events, then it means that several hours worth of block proposal and timeout events have been queued but not processed by server.go. This would mean that server.go is very broken, and that the node has more issues that need to be resolved in server.go aside from this deadlock